### PR TITLE
restXML Protocol Serializers

### PR DESCRIPTION
**Pull request by rcoh**
*Issue #, if available:* #114 
*Description of changes:* Need to add some more tests & cleanups, but restXML protocol tests are passing!

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/src/test/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parsers/XmlBindingTraitSerializerGeneratorTest.kt [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        }
+
+        project.withModule(RustModule.default("input", public = true)) {
+            model.lookup<OperationShape>("test#Op").inputShape(model).renderWithModelBuilder(model, symbolProvider, it)
+        }
+        println("file:///${project.baseDir}/src/xml_ser.rs")
+        try {
+            project.compileAndTest()
+        } finally {
+            // "cargo fmt".runCommand(project.baseDir)
```
Comment (jdisanti) [inline codegen/src/test/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parsers/XmlBindingTraitSerializerGeneratorTest.kt:149]: Is this left in to make debugging easier when the test fails? Just uncomment, rerun, and be able to read the code easier?

File: rust-runtime/protocol-test-helpers/src/xml.rs [hunk 2]
```diff
@@ -100,8 +100,13 @@ fn unparse_start_element(n: Node) -> String {
     let mut out = String::new();
     out.push('<');
     out.push_str(n.tag_name().name());
-    if let Some(ns) = n.tag_name().namespace() {
-        write!(&mut out, " xmlns=\"{}\"", ns).unwrap();
+    println!("{:?}", n);
```
Comment (jdisanti) [inline rust-runtime/protocol-test-helpers/src/xml.rs:103]: Left in intentionally?

Comment (rcoh) [reply rust-runtime/protocol-test-helpers/src/xml.rs:103]: definitely not, good spot

File: codegen/src/test/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parsers/XmlBindingTraitSerializerGeneratorTest.kt (hunk 1 above)
Comment (rcoh) [reply codegen/src/test/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parsers/XmlBindingTraitSerializerGeneratorTest.kt:148]: migrated this into `compileAndTest()` where it should have been all along

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parsers/XmlBindingTraitSerializerGenerator.kt [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+    sealed class Ctx {
+        abstract val input: String
+
+        data class Element(val elementWriter: String, override val input: String) : Ctx()
+        data class Scope(val scopeWriter: String, override val input: String) : Ctx()
+
+        companion object {
+            // Kotlin doesn't have a "This" type
+            @Suppress("UNCHECKED_CAST")
+            fun <T : Ctx> updateInput(inp: T, newInput: String): T = when (inp) {
```
[the remark this answers is not in the document]
Comment (rcoh) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parsers/XmlBindingTraitSerializerGenerator.kt:79]: yeah fine with me

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parsers/XmlBindingTraitSerializerGenerator.kt [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+    private val symbolProvider = protocolConfig.symbolProvider
+    private val runtimeConfig = protocolConfig.runtimeConfig
+    private val model = protocolConfig.model
+    private val smithyXml = CargoDependency.smithyXml(runtimeConfig).asType()
+    private val codegenScope =
+        arrayOf(
+            "XmlWriter" to smithyXml.member("encode::XmlWriter"),
+            "ElementWriter" to smithyXml.member("encode::ElWriter"),
+            "SdkBody" to RuntimeType.sdkBody(runtimeConfig),
+            // TODO: currently this doesn't every actually fail, however, once unions have an unknown member
```
Comment (jdisanti) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parsers/XmlBindingTraitSerializerGenerator.kt:63]:
```suggestion
            // TODO: currently this doesn't ever actually fail, however, once unions have an unknown member
```

**Review by jdisanti — APPROVED** [review]

**Pull request merged**